src/main/java/org/opentripplanner/routing/graphfinder/PlaceFinderTraverseVisitor.java

```diff
@@ -168,16 +171,32 @@ private boolean stopHasPatternsWithMode(RegularStop stop, Set<TransitMode> modes
   }
 
   private void handleStop(RegularStop stop, double distance) {
-    if (filterByStops != null && !filterByStops.contains(stop.getId())) {
+    if (
+      filterByStops != null &&
+      !filterByStops.contains(stop.getId()) ||
+      seenStops.contains(stop.getStationOrStopId()) ||
+      !(filterByModes == null || stopHasPatternsWithMode(stop, filterByModes))
+    ) {
```

Comment:
I know that you didn't write these but `filterByStops` and `filterByModes` should never be null. If they are then we should make sure in the constructor that they are not.